chain/network/src/routing.rs

```diff
@@ -855,6 +857,176 @@ impl Graph {
     }
 }
 
+ */
+
+#[derive(Clone)]
+pub struct Graph {
+    pub source: PeerId,
+    p2id: HashMap<PeerId, u64>,
+    id2p: HashMap<u64, PeerId>,
+    adjacency: HashMap<u64, HashSet<u64>>,
+    total_active_edges: u64,
+    nodes: u64,
```

Comment:
```suggestion
    id2p: Vec<PeerId>,
    adjacency: HashMap<u64, HashSet<u64>>,
    total_active_edges: u64,
```

This seems to be a more minimal representation -- `nodes == id2p.len()`, and looks like we don't actually need a hash map as ids are sequential?

Reply:
Nodes may get added and removed. The current code may leek memory in id2p, p2id if all edges of given code are removed. This code is just a proof of concept to show that performance can be optimized. This was not meant to be reviewed yet, as it's a draft.

Reply:
I'll have to add reference counting for Peers to know when they get removed. Once I do that, I can add another data structure with unused ids, and then I would be able to id2p to vector.